### DISCUSSION TITLE
Failed to connect to {}: Transport error: builder error: relative URL without a base

### DISCUSSION THREAD

**Discussion by Lykos153** [stalwartlabs/imap-server#2] [Q&A]
Hi,

I'm currently setting up both JMAP server and IMAP server in a k8s cluster. I can reach the JMAP server on `https://jmap.test.myserver.org`. But when I try to authenticate against the IMAP server, I get:

`DEBUG stalwart_imap::commands::authenticate: Failed to connect to {}: Transport error: builder error: relative URL without a base`

`{}` is the `jmap-url`, for which I tried all of the following with the same result:
`"https://jmap.test.myserver.org"`
`"jmap.test.myserver.org"`
`"https://jmap.test.myserver.org:443"`
`"jmap.test.myserver.org:443"`
`"https://jmap.test.myserver.org/.well-known/jmap"`

Does anybody have any idea what could be wrong?

Cheers
Silvio

**Comment by mdecimus** [comment]
Hi Silvio,

The correct format of the ``jmap-url`` parameter in your case should be ``https://jmap.test.myserver.org``. Is your IMAP k8s instance able to reach https://jmap.test.myserver.org? If it is then the problem might be that the JMAP server is configured with a URL that is either invalid or unreachable by the IMAP instance. Make sure that the ``jmap-url`` [parameter](https://stalw.art/jmap/configure/webserver/#jmap-url) on the JMAP server is a valid URL that is reachable by the IMAP instance.

-Mauro

**Reply by Lykos153** [reply]
Thanks for the reply! I double checked that `jmap-url` is `https://jmap.test.myserver.org` and that the IMAP server can reach the JMAP server under this url. I also tried to run the IMAP server outside the cluster to exclude any k8s internal side effects, but to no avail. The error stays the same. When I authenticate against `https://jmap.test.myserver.org/.well-known/jmap` using a browser, I can see that eg. `apiUrl` is set to `https://jmap.test.myserver.org/jmap/`, which seems right.

Now I'm thinking that my JMAP server setup must be broken in some other way. I actually couldn't get it to work with any of the clients that I tried. Though that may also be due to their immaturity. I could authenticate but then couldn't do anything (eg. sending mails to myself or to a different test account on the same server). I set the verbosity of the JMAP server to debug, but it doesn't say anything.

What client do you use to test and debug your server?

**Reply by mdecimus** [reply]
> The error stays the same. When I authenticate against https://jmap.test.myserver.org/.well-known/jmap using a browser, I can see that eg. apiUrl is set to https://jmap.test.myserver.org/jmap/, which seems right.

Could it be that your JMAP server is behind a reverse proxy that is redirecting the requests? Because based on the error message it would seem that the IMAP servers's HTTP client is trying to reach a relative path such as "/jmap" and not a full URL.
Another test you could do is setting up your IMAP server to use the demo JMAP instance running at ``https://jmap.cloud:8443`` using ``demo@stalw.art`` with password ``demo``.

> What client do you use to test and debug your server?

I use [Tmail](https://github.com/linagora/tmail-flutter) on the browser and [Ltt.rs](https://github.com/iNPUTmice/lttrs-android) on Android but both are a bit buggy unfortunately. Tmail has a problem with EmailSubmissions (which the developers said they have fixed last week but I haven't tried it yet) and message caching (when using multiple devices, messages deleted on device A still appear on device B). Ltt.rs works well overall but has some problems displaying message threads, but nothing critical.
However, for testing any of those should be fine. If you want to check Tmail out, it is running at ``https://jmap.cloud`` and you can use the login credentials I included above.
